### PR TITLE
CRM-16188, added order api for delete action and test

### PR DESCRIPTION
**Pull request by pradpnayak**
---
- CRM-16188: Create an order API
  https://issues.civicrm.org/jira/browse/CRM-16188

### CODE REVIEW

File: api/v3/Order.php [hunk 1]
```diff
@@ -59,3 +59,42 @@ function civicrm_api3_order_get($params) {
...
+/**
+ * Delete a Order.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @return array
+ */
+function civicrm_api3_order_delete($params) {
+  $contribution = civicrm_api3('Contribution', 'get', array(
```
Comment (eileenmcnaughton) [inline api/v3/Order.php]: Just a note here - one way to make this shorter would be 

if (!civicrm_api3('Contribution', 'getvalue', array('return' => 'is_test', 'id' => $params['id')) {
throw new ....
}

Note you can be sure $params['id'] will be set due to your aliasing - you don't need to check contribution_id.

Also - I suggest you edit the exception to say only test orders can be deleted.

File: api/v3/Order.php [hunk 2]
```diff
@@ -59,3 +59,42 @@ function civicrm_api3_order_get($params) {
...
+    'return' => array('is_test'),
+    'id' => CRM_Utils_Array::value('contribution_id', $params, $params['id']),
+  ));
+  if ($contribution['id'] && $contribution['values'][$contribution['id']]['is_test'] == TRUE) {
+    $result = civicrm_api3('Contribution', 'delete', $params);
+  }
+  else {
+    throw new API_Exception('Could not delete Order.');
+  }
+  return civicrm_api3_create_success(CRM_Utils_Array::value('values', $result), $params, 'Order', 'delete');
```
Comment (eileenmcnaughton) [inline api/v3/Order.php]: CRM_Utils_Array::value('values', $result) shouldn't be needed - $result['values'] should always be set & if it's not that is  a problem & an e-notice is appropriate

File: tests/phpunit/api/v3/OrderTest.php [hunk 3]
```diff
@@ -171,4 +171,27 @@ public function addOrder($isPriceSet, $amount = 300, $extraParams = array()) {
     return $this->callAPISuccess('Contribution', 'create', $p);
   }
 
+  /**
+   * Test delete order api
+   */
+  public function testDeleteOrder() {
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/OrderTest.php]: Nice test!

File: tests/phpunit/api/v3/OrderTest.php [hunk 4]
```diff
@@ -171,4 +171,27 @@ public function addOrder($isPriceSet, $amount = 300, $extraParams = array()) {
...
+    $order = $this->addOrder(FALSE, 100);
+    $params = array(
+      'contribution_id' => $order['id'],
+    );
+    try {
+      $this->callAPISuccess('order', 'delete', $params);
+      $this->fail("Missed expected exception");
+    }
+    catch (Exception $expected) {
+      $this->callAPISuccess('contribution', 'create', array(
```
Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/OrderTest.php]: A trivial point - but we are trying to capitalise the entity as a convention & lower case the action - @colemanw  may have documented that somewhere?
ie
$this->callAPISuccess('Contribution', 'create', array(

**Comment by eileenmcnaughton** [comment]
The test fails are unrelated but I think it's worth at minimum expanding the api exception message to state the reason and getting rid of the unnecessary e-notice preventions. The other comments are more in the nice-to-have.

I don't see any larger concerns with this PR

File: api/v3/Order.php (hunk 1 above)
Comment (pradpnayak) [inline api/v3/Order.php]: I had tried with 

  $result = civicrm_api3('Contribution', 'getvalue', array(
      'sequential' => 1,
      'return' => array("is_test"),
      'id' => 22,
    ));

I was getting exception error with 'field Array unset or not existing'.

Comment (eileenmcnaughton) [inline api/v3/Order.php]: ah I think that might be a bug in the contribution api - no accepting that as a return value

**Comment by eileenmcnaughton** [comment]
I've updated the title to [needs minor changes] to reflect my comment above - please change to [needs core team review] when you have fixed the exception (& preferably the enotice over-handling)

**Pull request merged**